### PR TITLE
feat(cli): add config management commands

### PR DESCRIPTION
**Pull request by JanakaSandaruwan**
## Purpose
resolves https://github.com/openchoreo/openchoreo/issues/1507

## Checklist
- [ ] Tests added or updated (unit, integration, etc.)
- [ ] Samples updated (if applicable)

# Pull Request Summary: CLI Config Command Restructuring

## Mini breakdown — changed files by top-level folder (from provided diffs)
- api/: 0 files changed  
- config/: 1 file changed  
- internal/: 3 files changed (internal/occ/cmd/config/config.go, internal/occ/cmd/login/login.go, internal/occ/impl.go) — aggregate approx. +432 / -186 lines  
- pkg/: 7 files changed (pkg/cli/cmd/config/config.go, pkg/cli/cmd/config/types.go, pkg/cli/cmd/login/login.go, pkg/cli/common/constants/definitions.go, pkg/cli/flags/flags.go, pkg/cli/types/api/api.go, pkg/cli/types/api/params.go) — aggregate approx. +309 / -150 lines  
- install/: 0 files changed  
- docs/: 0 files changed  
- openapi/: 0 files changed  
- rca-agent/: 0 files changed  
- make/: 0 files changed  
- cmd/: 0 files changed  
- samples/: 0 files changed

Total files changed reported: 10 — net approx. +741 / -336 lines (sum across provided diffs)

## API / Public surface changes (compatibility risk: HIGH)
- CLI command surface restructured from flat commands to nested groups:
  - Removed top-level commands: config get-contexts, config set-context, config use-context, config current-context, config set-control-plane.
  - Added nested subcommands:
    - config context: add, list, delete, update, use/describe
    - config control-plane: add, list, update, delete
    - config credentials: add, list, delete
  - New flags: --controlplane, --credentials
- Public Go API changes (pkg/cli/types/api and internal/occ):
  - Removed/renamed methods: GetContexts(), GetCurrentContext(), SetContext(...), SetControlPlane(...) removed.
  - Added CRUD-style methods: AddContext/ListContexts/DeleteContext/UpdateContext/DescribeContext; AddControlPlane/ListControlPlanes/UpdateControlPlane/DeleteControlPlane; AddCredentials/ListCredentials/DeleteCredentials.
  - Params types split/renamed (AddContextParams, UpdateContextParams, DescribeContextParams, AddControlPlaneParams, etc.); LoginParams no longer contains URL.
- New exported types (pkg/cli/cmd/config/types.go): StoredConfig, ControlPlane, Credential, Context, and Mode constants.
- Compatibility risk: HIGH — breaking changes to CLI command names/flags and public API signatures/types; downstream code, automation, and docs will require updates.

## Tests added/updated
- No test files were reported as added or modified in the provided summaries (no *_test.go changes).
- Critical new/changed paths currently lacking tests:
  - Context CRUD flows (Add/List/Update/Delete/Describe), including name-uniqueness validation and auto-creation behavior.
  - ControlPlane CRUD flows, including referential constraints preventing deletion when referenced.
  - Credentials CRUD flows, persistence, masking and secure handling.
  - validateNameUniqueness and new validation helpers (ValidateAddContextParams, Validate*NameUniqueness).
  - StoredConfig schema changes and migration behavior (DataPlane/Environment removed).
  - Login behavior change (removal of URL propagation) and interactions with pre-configured control-planes.

## Risk hotspots (short reasons)
1. Breaking API / CLI surface (HIGH) — renamed/removed public methods and CLI commands/flags will break consumers, CI scripts, and docs.  
2. Config storage & migration (HIGH) — StoredConfig schema changed (new types, removed fields). No migration logic or tests provided; risk of load failures or data loss.  
3. Authentication / control-plane linkage (MEDIUM) — login no longer accepts/persists a URL; workflows expecting login to set control-plane URL will be affected.  
4. Secrets & credentials handling (MEDIUM) — credentials are persisted and displayed (masking added) but secure storage/migration/rotation considerations and tests are missing.  
5. Referential integrity & concurrency (MEDIUM) — deletion protections and auto-creation (AddContext can create ControlPlane/Credentials) require tests to avoid orphaned entries or race conditions.  
6. UX / discoverability (LOW–MEDIUM) — nested commands improve organization but will break existing examples and require documentation updates.

## Recommendations / action items
- Add unit and integration tests covering:
  - All new CLI handlers (context/control-plane/credentials): success/failure/edge cases, persistence, masking, and referential constraints.
  - Validation helpers and uniqueness checks.
  - Migration tests for existing config files and StoredConfig schema changes.
  - Login flow tests verifying absence of URL propagation and interactions with stored control-planes.
- Publish a migration/compatibility guide mapping old commands/flags and API signatures to the new ones; update user docs and examples.
- Add CI checks to exercise CLI help/usage text and examples to catch doc regressions.

### CODE REVIEW

File: internal/occ/cmd/config/config.go [hunk 1]
```diff
@@ -23,8 +22,67 @@ func NewConfigContextImpl() *ConfigContextImpl {
...
+	if err != nil {
+		return fmt.Errorf("failed to load config: %w", err)
+	}
+
+	// Validate that the name is not already used by a context, control plane, or credential
+	if err := validateNameUniqueness(cfg, params.Name); err != nil {
+		return err
+	}
+
+	// Create control plane entry if it doesn't exist
```
Comment (binoyPeries) [inline internal/occ/cmd/config/config.go:44]: what does this do exactly?

File: internal/occ/cmd/config/config.go [hunk 2]
```diff
@@ -35,198 +93,112 @@ func (c *ConfigContextImpl) GetContexts() error {
 		return nil
 	}
 
-	// Create headers and rows for table
-	headers := []string{"", "NAME", "NAMESPACE", "PROJECT", "COMPONENT", "ENVIRONMENT", "DATAPLANE"}
+	headers := []string{"", "NAME", "CONTROLPLANE", "CREDENTIALS", "NAMESPACE", "PROJECT", "COMPONENT"}
```
Comment (binoyPeries) [inline internal/occ/cmd/config/config.go:112]: please add comment saying why there's an empty header at the beginning

File: internal/occ/cmd/config/config.go [hunk 3]
```diff
@@ -23,8 +22,67 @@ func NewConfigContextImpl() *ConfigContextImpl {
...
+	}
+	cfg.Contexts = append(cfg.Contexts, newCtx)
+
+	if err := SaveStoredConfig(cfg); err != nil {
+		return fmt.Errorf("failed to save config: %w", err)
+	}
+
+	fmt.Printf("Created context: %s\n", params.Name)
+	return nil
+}
```
[the remark this answers is not in the document]
Comment (JanakaSandaruwan) [reply internal/occ/cmd/config/config.go:97]: Fixed with https://github.com/openchoreo/openchoreo/pull/1878/changes/f6535aa2d3e289c5807b2c765442de0692c4cc9f

File: internal/occ/cmd/config/config.go [hunk 4]
```diff
@@ -404,53 +428,111 @@ func EnsureContext() error {
...
+		URL:  params.URL,
+	})
+
+	if err := SaveStoredConfig(cfg); err != nil {
+		return fmt.Errorf("failed to save config: %w", err)
+	}
+
+	fmt.Printf("Created control plane: %s\n", params.Name)
+	return nil
+}
```
Comment (JanakaSandaruwan) [inline internal/occ/cmd/config/config.go:493]: Fixed with https://github.com/openchoreo/openchoreo/pull/1878/changes/f6535aa2d3e289c5807b2c765442de0692c4cc9f

File: internal/occ/cmd/config/config.go [hunk 5]
```diff
@@ -404,53 +428,111 @@ func EnsureContext() error {
...
-	fmt.Printf("Control plane configured successfully:\n")
-	fmt.Printf("  Name: %s\n", params.Name)
-	fmt.Printf("  URL: %s\n", params.URL)
+	if err := SaveStoredConfig(cfg); err != nil {
+		return fmt.Errorf("failed to save config: %w", err)
+	}
 
+	fmt.Printf("Deleted control plane: %s\n", params.Name)
 	return nil
 }
```
Comment (JanakaSandaruwan) [inline internal/occ/cmd/config/config.go:592]: Fixed with https://github.com/openchoreo/openchoreo/pull/1878/changes/f6535aa2d3e289c5807b2c765442de0692c4cc9f

File: internal/occ/cmd/config/config.go [hunk 6]
```diff
@@ -23,8 +23,83 @@ func NewConfigContextImpl() *ConfigContextImpl {
...
+	}
+	if !cpExists {
+		// Validate uniqueness before creating
+		if err := validateNameUniqueness(cfg, params.ControlPlane); err != nil {
+			return fmt.Errorf("cannot create control plane: %w", err)
+		}
+		cfg.ControlPlanes = append(cfg.ControlPlanes, configContext.ControlPlane{
+			Name: params.ControlPlane,
+		})
+	}
```
[the remark this answers is not in the document]
Comment (JanakaSandaruwan) [reply internal/occ/cmd/config/config.go:60]: The expected behaviour is to set url seperately

File: internal/occ/cmd/config/config.go [hunk 7]
```diff
@@ -404,53 +445,124 @@ func EnsureContext() error {
...
-	for idx := range cfg.Contexts {
-		if cfg.Contexts[idx].Name == currentContext.Name {
-			cfg.Contexts[idx].ControlPlane = params.Name
+	if err := SaveStoredConfig(cfg); err != nil {
+		return fmt.Errorf("failed to save config: %w", err)
+	}
+
+	fmt.Printf("Updated control plane: %s\n", params.Name)
+	return nil
+}
```
[the remark this answers is not in the document]
Comment (JanakaSandaruwan) [reply internal/occ/cmd/config/config.go:557]: fixed with 8174b08fa961da499c59543054b1457020b2041b

File: internal/occ/cmd/config/config.go (hunk 1 above)
Comment (JanakaSandaruwan) [reply internal/occ/cmd/config/config.go:44]: This adds a new entry to the control plane section in config file

File: internal/occ/cmd/config/config.go (hunk 2 above)
Comment (JanakaSandaruwan) [reply internal/occ/cmd/config/config.go:112]: First empty column for current context marker. Added a comment

File: internal/occ/cmd/config/config.go [hunk 8]
```diff
@@ -404,53 +445,133 @@ func EnsureContext() error {
...
+		URL:  trimmedURL,
+	})
+
+	if err := SaveStoredConfig(cfg); err != nil {
+		return fmt.Errorf("failed to save config: %w", err)
+	}
+
+	fmt.Printf("Created control plane: %s\n", params.Name)
+	return nil
+}
```
[the remark this answers is not in the document]
Comment (JanakaSandaruwan) [reply internal/occ/cmd/config/config.go:493]: This is already validated at newControlPlaneAddCmd

File: internal/occ/cmd/config/config.go [hunk 9]
```diff
@@ -23,8 +23,83 @@ func NewConfigContextImpl() *ConfigContextImpl {
...
+		Component:    params.Component,
+	}
+	cfg.Contexts = append(cfg.Contexts, newCtx)
+
+	if err := SaveStoredConfig(cfg); err != nil {
+		return fmt.Errorf("failed to save config: %w", err)
+	}
+
+	fmt.Printf("Created context: %s\n", params.Name)
+	return nil
```
[the remark this answers is not in the document]
Comment (JanakaSandaruwan) [reply internal/occ/cmd/config/config.go:96]: This is already validated at newContextAddCmd

File: pkg/cli/cmd/config/config.go [hunk 10]
```diff
@@ -134,40 +156,169 @@ func newUseContextCmd(impl api.CommandImplementationInterface) *cobra.Command {
...
+		},
+	}).Build()
+
+	cmd.Args = cobra.ExactArgs(1)
 
-			return impl.SetControlPlane(api.SetControlPlaneParams{
-				Name: name,
-				URL:  url,
+	return cmd
+}
```
[the remark this answers is not in the document]
Comment (JanakaSandaruwan) [reply pkg/cli/cmd/config/config.go:292]: Yes. Users have to run the login command to set the token and other info to the current context's credentials. Once the PAT support is added this will get --token as well

File: pkg/cli/types/api/api.go [hunk 11]
```diff
@@ -85,11 +85,19 @@ type ObservabilityPlaneAPI interface {
...
-	SetContext(params SetContextParams) error
-	SetControlPlane(params SetControlPlaneParams) error
+	DescribeContext(params DescribeContextParams) error
+	AddControlPlane(params AddControlPlaneParams) error
+	ListControlPlanes() error
+	UpdateControlPlane(params UpdateControlPlaneParams) error
+	DeleteControlPlane(params DeleteControlPlaneParams) error
+	AddCredentials(params AddCredentialsParams) error
+	ListCredentials() error
+	DeleteCredentials(params DeleteCredentialsParams) error
```
[the remark this answers is not in the document]
Comment (JanakaSandaruwan) [reply pkg/cli/types/api/api.go:100]: Yes. This is a breaking change to align with new occ command structure

**Review by binoyPeries — APPROVED** [review]
